File: 55-jump-game/55.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        valid = [False] * len(nums)
        valid[0] = True

        for i in range(len(nums)):
            if valid[i] == False:
                continue
            if i == 0:
                start = 1
            else:
                if nums[i] < nums[i-1]:
                    continue
                start = i + nums[i-1]
            for j in range(start, i + nums[i] + 1):
                if j < len(nums):
                    if j == len(nums) - 1:
                        return True
                    valid[j] = True
                else:
                    return True

        return valid[-1]

File: 55-jump-game/55.c
#define max(a,b) ((a) > (b) ? (a) : (b))

bool canJump(int* nums, int numsSize){
    int dp[2] = {0, 0};

    for (int i = 1; i < numsSize; ++i) {
        dp[i&1] = max(dp[i-1&1], nums[i-1]) - 1;
        if (dp[i&1] < 0)
            return false;
    }
    return true;
}
